rmw_opendds_cpp/src/rmw_take.cpp

```diff
@@ -26,179 +26,114 @@
-    DDS::InstanceHandle_t receiver_instance_handle = dds_data_reader->get_instance_handle();
-    ignore_sample = true;
-    for (size_t i = 0; i < 12; ++i) {
-      //CORBA::Octet * sender_element = &(sender_guid.value[i]);
-      CORBA::Octet * receiver_element =
-        &(reinterpret_cast<CORBA::Octet *>(&receiver_instance_handle)[i]);
-      //if (*sender_element != *receiver_element) {
-      //  ignore_sample = false;
-      //  break;
-      //}
+  DDS::ReturnCode_t status = reader->take(dds_messages, sample_infos, 1,
+    DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ANY_INSTANCE_STATE);
+
+  if (DDS::RETCODE_OK == status) {
+    DDS::SampleInfo & info = sample_infos[0];
+    if (info.valid_data) {
+      cdr_stream->buffer_length = dds_messages[0].serialized_data.length();
+      if (cdr_stream->buffer_length <= (std::numeric_limits<unsigned int>::max)()) {
+        cdr_stream->buffer = reinterpret_cast<uint8_t *>(malloc(cdr_stream->buffer_length * sizeof(uint8_t)));
+        for (unsigned int i = 0; i < static_cast<unsigned int>(cdr_stream->buffer_length); ++i) {
+          cdr_stream->buffer[i] = dds_messages[0].serialized_data[i];
+        }
+        *taken = true;
+
+        if (message_info) {
+          message_info->publisher_gid.implementation_identifier = opendds_identifier;
+          memset(message_info->publisher_gid.data, 0, RMW_GID_STORAGE_SIZE);
+          auto detail = reinterpret_cast<OpenDDSPublisherGID *>(message_info->publisher_gid.data);
+          detail->publication_handle = info.publication_handle;
+        }
+      } else {
+        RMW_SET_ERROR_MSG("cdr_stream->buffer_length > max unsigned int");
+      }
     }
+  } else if (DDS::RETCODE_NO_DATA != status) {
+    RMW_SET_ERROR_MSG("take failed");
   }
-  if (sample_info.valid_data && sending_publication_handle) {
-    *static_cast<DDS::InstanceHandle_t *>(sending_publication_handle) =
-      sample_info.publication_handle;
-  }
-
-  if (!ignore_sample) {
-    cdr_stream->buffer_length = dds_messages[0].serialized_data.length();
-    // TODO(karsten1987): This malloc has to go!
-    cdr_stream->buffer =
-      reinterpret_cast<uint8_t *>(malloc(cdr_stream->buffer_length * sizeof(uint8_t)));
-
-    if (cdr_stream->buffer_length > (std::numeric_limits<unsigned int>::max)()) {
-      RMW_SET_ERROR_MSG("cdr_stream->buffer_length unexpectedly larger than max unsiged int value");
-      data_reader->return_loan(dds_messages, sample_infos);
-      *taken = false;
-      return false;
-    }
-    for (unsigned int i = 0; i < static_cast<unsigned int>(cdr_stream->buffer_length); ++i) {
-      cdr_stream->buffer[i] = dds_messages[0].serialized_data[i];
-    }
-    *taken = true;
-  } else {
-    *taken = false;
-  }
-
-  data_reader->return_loan(dds_messages, sample_infos);
 
-  return status == DDS::RETCODE_OK;
+  reader->return_loan(dds_messages, sample_infos);
+  return (*taken || DDS::RETCODE_NO_DATA == status) ? RMW_RET_OK : RMW_RET_ERROR;
```

Comment:
Before this change, wasn't NO_DATA an error?  Is this something we have to come back to later, or a deliberate change?

Reply:
This is not a change of logic, nor a temporary code. Whether I implemented it correctly needs your verification. This c-style function has more than one values for the caller. The original code returns a bool:

  if (status == DDS::RETCODE_NO_DATA) {
    data_reader->return_loan(dds_messages, sample_infos);
    *taken = false;
    return true;
  }
  if (status != DDS::RETCODE_OK) {
    RMW_SET_ERROR_MSG("take failed");
    data_reader->return_loan(dds_messages, sample_infos);
    return false;
  }

I think *taken tells the caller whether to deserialize and then free the stream buffer. So NO_DATA is not an error, but the caller cannot process the stream. I factored out this logic from all the callers for simplicity and consistency. Its correctness needs your verification.

Reply:
OK, it must have just been hard to see from the diff comparison.